### DISCUSSION TITLE
Ubuntu 21.10 will ship firefox as Snap

### DISCUSSION THREAD

**Discussion by rusty-snake** [netblue30/firejail#4554] [General]
I am already excited about the issues :innocent:

https://bugs.launchpad.net/ubuntu/+source/ubuntu-release-upgrader/+bug/1943840
https://discourse.ubuntu.com/t/feature-freeze-exception-seeding-the-official-firefox-snap-in-ubuntu-desktop/24210

[Edit by @reinerh]: Fixed link

**Comment by kmk3** [comment]
https://discourse.ubuntu.com/t/feature-freeze-exception-seeding-the-official-firefox-snap-in-ubuntu-desktop/24210l

> replace the Firefox seed in ubuntu-desktop from a deb package to a
> Mozilla-official Firefox snap

Well, Ubuntu wasn't exactly on my recommendation list before and now it's in
the "actively avoid" list.  Hopefully the derivative distros (e.g.: Linux Mint)
can keep up working around the damage.

**Comment by slowpeek** [comment]
> Mozilla-official Firefox snap

Even though there are Mozilla-official Firefox builds for linux (no armhf though).

>  Hopefully the derivative distros (e.g.: Linux Mint)

Mint said they'll throw the snap out of the window since the time it was announced for the first time.

**Comment by Keyshoun** [comment]
I use Mozilla-official Firefox builds for linux now, but they cause trouble with firejail too. Are there any profiles that are ready for these builds with custom path of the firefox binary?

It is not enough to put the custom path as whitelist in the firejail firefox.profile.

Yet i can only run firefox with --noprofile :(

Starting it with firefox.profile is ending with: firejail no suitable executable found

**Reply by reinerh** [reply]
What are Mozilla-official Firefox builds? Do you mean binary archives you just unpack to /opt or so?

If you use some Ubuntu-derived distribution you can also use a PPA (maintained by Mozilla team): https://balintreczey.hu/blog/firefox-on-ubuntu-22-04-from-deb-not-from-snap/

**Reply by rusty-snake** [reply]
> Are there any profiles that are ready for these builds with custom path of the firefox binary?

No, because custom paths are ... custom i.e. can be anywhere.

You can try with this `firefox.local`:

```
whitelist ${HOME}/.local/opt/mozilla/firefox
ignore noexec ${HOME}
ignore apparmor
```

>  Mozilla-official Firefox builds

Unfortunately the official builds (flatpak and snap(?) are official builds too) lack PIE and Full RELRO while distro builds - at least fedora - has them enabled.

**Reply by Keyshoun** [reply]
> What are Mozilla-official Firefox builds? Do you mean binary archives you just unpack to /opt or so?
> 
> If you use some Ubuntu-derived distribution you can also use a PPA (maintained by Mozilla team): https://balintreczey.hu/blog/firefox-on-ubuntu-22-04-from-deb-not-from-snap/

I know about these repo already, but in general i prefer the Mozilla-official Firefox binary builds. Yes, exactly these builds i am talking about. You can extract them to any path and just start firefox from there.



> > Are there any profiles that are ready for these builds with custom path of the firefox binary?
> 
> No, because custom paths are ... custom i.e. can be anywhere.
> 
> You can try with this `firefox.local`:
> 
> ```
> whitelist ${HOME}/.local/opt/mozilla/firefox
> ignore noexec ${HOME}
> ignore apparmor
> ```
> 
> > Mozilla-official Firefox builds
> 
> Unfortunately the official builds (flatpak and snap(?) are official builds too) lack PIE and Full RELRO while distro builds - at least fedora - has them enabled.

My path is not in the ${HOME} path it is something like /apps/browser/firefox is that complicating things?
This is how i would run it, but it seems these changes are not enough.
firejail --ignore=apparmor --ignore='noexec /apps/browser/firefox' --whitelist=/apps/browser/firefox --profile=/etc/firejail/firefox.profile /apps/browser/firefox/firefox

**Reply by rusty-snake** [reply]
> My path is not in the ${HOME} path it is something like /apps/browser/firefox is that complicating things?

Actaully it should make things easier.

- The `ignore noexec ...` can be skiped.
- The `whitelist` can be done as hardening but is not required so leave it for now until we have found a way to start it.
- About `apprmor` I'm unsure to ignore it for now.
- Do you have an included .locals?
- Do you have any kind of symlink going to `/apps...`?
- What does `firejail --ignore=apparmor --profile=firefox /usr/bin/ls /apps/browser/firefox` show?

**Reply by Keyshoun** [reply]
> > My path is not in the ${HOME} path it is something like /apps/browser/firefox is that complicating things?
> 
> Actaully it should make things easier.
> 
> * The `ignore noexec ...` can be skiped.
> * The `whitelist` can be done as hardening but is not required so leave it for now until we have found a way to start it.
> * About `apprmor` I'm unsure to ignore it for now.
> * Do you have an included .locals?
> * Do you have any kind of symlink going to `/apps...`?
> * What does `firejail --ignore=apparmor --profile=firefox /usr/bin/ls /apps/browser/firefox` show?

I think i was not in contact with any .locals yet. Are they part of firejail profiles?

Yes, i have a symlink /usr/bin/firefox pointing to /apps/browser/firefox/firefox

**Reply by rusty-snake** [reply]
https://github.com/netblue30/firejail/wiki/Creating-overrides

**Reply by Keyshoun** [reply]
> * What does `firejail --ignore=apparmor --profile=firefox /usr/bin/ls /apps/browser/firefox` show?

Result:
/usr/bin/ls: cannot access '/apps/browser/firefox': Permission denied

**Reply by rusty-snake** [reply]
Now the question why.

Does `/apps` and `/apps/browser` work?

Any other symlink to `/apps` or `/apps/browser` or `/apps/browser/firefox`?

Who owns `/apps`, `/apps/browser`, `/apps/browser/firefox` with which permissions?

**Reply by Keyshoun** [reply]
> https://github.com/netblue30/firejail/wiki/Creating-overrides

ahh i think that is what i was looking for. I don`t wanted to change the firefox.profile that`s why i used the command line flags so far for testing.

**Reply by Keyshoun** [reply]
> Now the question why.
> 
> Does `/apps` and `/apps/browser` work?
> 
> Any other symlink to `/apps` or `/apps/browser` or `/apps/browser/firefox`?
> 
> Who owns `/apps`, `/apps/browser`, `/apps/browser/firefox` with which permissions?

No also not working, but i can access all the dirs with my local user, no root needed and also the firefox is starting from there when i use --noprofile option, so it should be not filesystem permission issue.

No that`s the only symlink i made for that path.

All owned by my local user

**Reply by rusty-snake** [reply]
So `firejail --ignore=apparmor --profile=firefox /usr/bin/ls -ld /apps` shows `000 nobody nobody`?

Is `/apps` an mount point? If so anything "special"?

**Reply by Keyshoun** [reply]
> So `firejail --ignore=apparmor --profile=firefox /usr/bin/ls -ld /apps` shows `000 nobody nobody`?
> 
> Is `/apps` an mount point? If so anything "special"?

it shows:
dr-------- 2 nobody nogroup

/apps is no mount point, just a folder in the root filesystem

**Reply by rusty-snake** [reply]
Either you added a `blacklist` for it, or you symlinked it from a path that has a `blacklist`.

Run with `--debug` ang go trough it (Ctrl-F) to find it.

**Comment by slowpeek** [comment]
I use Mozilla official builds with firejail no problem.

The extracted archive goes into `/opt/mozilla/` so the path to the binary is `/opt/mozilla/firefox/firefox`.
`/usr/bin/firefox` is a symlink to `/opt/mozilla/firefox/firefox`.
`/usr/local/bin/firefox` is a symlink to `/usr/bin/firejail`.

And here is my desktop file for firefox:
```
[Desktop Entry]
Version=1.0
Type=Application
Terminal=false
StartupNotify=true
Name=Firefox
Exec=firefox %u
Icon=/opt/mozilla/firefox/browser/chrome/icons/default/default128.png

Categories=Network;WebBrowser;
MimeType=text/html;text/xml;application/xhtml+xml;application/xml;application/rss+xml;application/rdf+xml;image/gif;image/jpeg;image/png;x-scheme-handler/http;x-scheme-handler/https;x-scheme-handler/chrome;video/webm;application/x-xpinstall;
Actions=new-window;new-private-window;

[Desktop Action new-window]
Name=Open a new window
Exec=firefox -new-window

[Desktop Action new-private-window]
Name=Open a new private window
Exec=firefox -private-window
```

Work as supposed with that. Same for thunderbird. Here its desktop file:
```
[Desktop Entry]
Version=1.0
Type=Application
Terminal=false
StartupNotify=true
Name=Thunderbird
Exec=thunderbird %u
Icon=/opt/mozilla/thunderbird/chrome/icons/default/default128.png

Categories=Network;Email;
MimeType=x-scheme-handler/mailto;application/x-xpinstall;
Actions=compose;contacts

[Desktop Action compose]
Name=Compose new message
Exec=thunderbird -compose

[Desktop Action contacts]
Name=Contacts
Exec=thunderbird -addressbook
```

When I need to update I just drop current `/opt/mozilla/firefox` and extract the new version there.

**Reply by rusty-snake** [reply]
How do you get notifications about new updates?

**Reply by slowpeek** [reply]
> How do you get notifications about new updates?

I set "Check for updates but let you choose to install them" in the settings. There is a popup when an update is available.

**Reply by Keyshoun** [reply]
> I use Mozilla official builds with firejail no problem.
> 
> The extracted archive goes into `/opt/mozilla/` so the path to the binary is `/opt/mozilla/firefox/firefox`. `/usr/bin/firefox` is a symlink to `/opt/mozilla/firefox/firefox`. `/usr/local/bin/firefox` is a symlink to `/usr/bin/firejail`.
> 
> And here is my desktop file for firefox:
> 
> ```
> [Desktop Entry]
> Version=1.0
> Type=Application
> Terminal=false
> StartupNotify=true
> Name=Firefox
> Exec=firefox %u
> Icon=/opt/mozilla/firefox/browser/chrome/icons/default/default128.png
> 
> Categories=Network;WebBrowser;
> MimeType=text/html;text/xml;application/xhtml+xml;application/xml;application/rss+xml;application/rdf+xml;image/gif;image/jpeg;image/png;x-scheme-handler/http;x-scheme-handler/https;x-scheme-handler/chrome;video/webm;application/x-xpinstall;
> Actions=new-window;new-private-window;
> 
> [Desktop Action new-window]
> Name=Open a new window
> Exec=firefox -new-window
> 
> [Desktop Action new-private-window]
> Name=Open a new private window
> Exec=firefox -private-window
> ```
> 
> Work as supposed with that. Same for thunderbird. Here its desktop file:
> 
> ```
> [Desktop Entry]
> Version=1.0
> Type=Application
> Terminal=false
> StartupNotify=true
> Name=Thunderbird
> Exec=thunderbird %u
> Icon=/opt/mozilla/thunderbird/chrome/icons/default/default128.png
> 
> Categories=Network;Email;
> MimeType=x-scheme-handler/mailto;application/x-xpinstall;
> Actions=compose;contacts
> 
> [Desktop Action compose]
> Name=Compose new message
> Exec=thunderbird -compose
> 
> [Desktop Action contacts]
> Name=Contacts
> Exec=thunderbird -addressbook
> ```
> 
> When I need to update I just drop current `/opt/mozilla/firefox` and extract the new version there.

This really works, thanks a lot.

**Reply by slowpeek** [reply]
> This really works, thanks a lot.

Glad to hear!
